### PR TITLE
Update snapshot-for-clones doc to reseed RNG

### PR DESCRIPTION
**Pull request by mattschlebusch**
## Changes

Following recent changes to the Linux Kernel, synchronous reseeding of entropy pools has been changed and does not necessarily guarantee secure random number generation depending on the timeframe (see [reseed interval for CRNG](https://elixir.bootlin.com/linux/v4.14.295/source/drivers/char/random.c#L217) after a restored snapshot has been resumed.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/snapshotting/random-for-clones.md [hunk 1]
```diff
@@ -114,26 +114,27 @@ the read result via bind mounting another file on top of
...
+        This should also cause the `/dev/urandom` `CSPRNG`
+        to be reseeded. The bytes can be generated locally in the guest,
+        or obtained from the host.
+     2. Take note that `RNDCLEARPOOL` no [longer has any effect][7] on
+        the entropy pool.
 
 **Annex 1 contains the source code of a C program which implements the
 previous three steps.** As soon as the guest kernel version switches to
-4.19 (or higher), we can rely on the `CONFIG_RANDOM_TRUST_CPU` kernel
+4.19 (or higher), we can rely on the `CONF5IG_RANDOM_TRUST_CPU` kernel
```
Comment (kalyazin) [inline docs/snapshotting/random-for-clones.md:137]: looks like typo to me

File: docs/snapshotting/random-for-clones.md [hunk 2]
```diff
@@ -114,22 +114,23 @@ the read result via bind mounting another file on top of
...
-  1. Issue an `RNDADDENTROPY ioctl` (requires `CAP_SYS_ADMIN`) to mix
-     the provided bytes into the input entropy pool and increase the
-     entropy count. This should also cause the `/dev/urandom` `CSPRNG`
-     to be reseeded. The bytes can be generated locally in the guest,
-     or obtained from the host. Starting with version 4.17 of the
-     kernel, there’s a new ioctl request (`RNDRESEEDCRNG`) that
+  2. Issue a `RNDRESEEDCRNG` system call.
+     `RNDRESEEDCRNG` ([4.14][5], [5.10][6], (requires `CAP_SYS_ADMIN`)) that
      specifically causes the `CSPRNG` to be reseeded from the input pool.
+     1. An alternative would be to issue an `RNDADDENTROPY ioctl`
```
Comment (bchalios) [inline docs/snapshotting/random-for-clones.md:126]: See above, not sure if this should be an "alternative" or mandatory step (as per initial recommendations). Do you have any more context here?

Comment (mattschlebusch) [reply docs/snapshotting/random-for-clones.md:126]: Yea I think you are spot on. I misunderstood the [crng_reseed()](https://elixir.bootlin.com/linux/v4.14.295/source/drivers/char/random.c#L669) call as accommodating entropy increase as well. the random.c file has been heavily refactored. I will revert on this specific point to make use of RNDADDENTROPY. The RNDADDENTROPY ioctl call doesn't seem to necessarily reseed though, would you agree that forcing a reseed would be prudent afterwards as well?

File: docs/snapshotting/random-for-clones.md [hunk 3]
```diff
@@ -114,22 +114,22 @@ the read result via bind mounting another file on top of
...
-     specifically causes the `CSPRNG` to be reseeded from the input pool.
+     `/dev/urandom`). Take note that `RNDCLEARPOOL` no longer
+     [has any effect][7] on the entropy pool.
+  1. Issue an `RNDADDENTROPY` ioctl call (requires `CAP_SYS_ADMIN`)
+   to mix the provided bytes into the input
+   entropy pool and increase the entropy count.
+   This should also cause the `/dev/urandom` `CSPRNG`
+   to be reseeded. The bytes can be generated locally in the guest,
+   or obtained from the host.
+  1. Issue a `RNDRESEEDCRNG` ioctl call
```
Comment (dianpopa) [inline docs/snapshotting/random-for-clones.md:130]: This was on 4.14 but the doc said that only 4.17 and above has it, am I getting this right?

Comment (mattschlebusch) [reply docs/snapshotting/random-for-clones.md:130]: The original commit to remove the call to `RDRAND` was made to kernel 5.17, however the commit has been backported to both [4.14](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-4.14.y&id=d15b7abb41918efbeadf27119d67bdfbcff422e3) and [5.10](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-5.10.y&id=d15b7abb41918efbeadf27119d67bdfbcff422e3).

File: docs/snapshotting/random-for-clones.md [hunk 4]
```diff
@@ -114,22 +114,22 @@ the read result via bind mounting another file on top of
...
-     cause the reinitialization of the `/dev/urandom` `CSPRNG`.
-  1. Issue an `RNDADDENTROPY ioctl` (requires `CAP_SYS_ADMIN`) to mix
-     the provided bytes into the input entropy pool and increase the
-     entropy count. This should also cause the `/dev/urandom` `CSPRNG`
-     to be reseeded. The bytes can be generated locally in the guest,
-     or obtained from the host. Starting with version 4.17 of the
-     kernel, there’s a new ioctl request (`RNDRESEEDCRNG`) that
-     specifically causes the `CSPRNG` to be reseeded from the input pool.
+     `/dev/urandom`). Take note that `RNDCLEARPOOL` no longer
+     [has any effect][7] on the entropy pool.
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply docs/snapshotting/random-for-clones.md:123]: It does indeed, [5.10](https://elixir.bootlin.com/linux/v5.10.148/source/drivers/char/random.c#L1365).

File: docs/snapshotting/random-for-clones.md [hunk 5]
```diff
@@ -114,22 +114,23 @@ the read result via bind mounting another file on top of
...
-  1. Issue an `RNDCLEARPOOL ioctl` (requires `CAP_SYS_ADMIN`). This
-     clears and sets the entropy pools to the initial state. Should also
-     cause the reinitialization of the `/dev/urandom` `CSPRNG`.
-  1. Issue an `RNDADDENTROPY ioctl` (requires `CAP_SYS_ADMIN`) to mix
-     the provided bytes into the input entropy pool and increase the
-     entropy count. This should also cause the `/dev/urandom` `CSPRNG`
-     to be reseeded. The bytes can be generated locally in the guest,
-     or obtained from the host. Starting with version 4.17 of the
-     kernel, there’s a new ioctl request (`RNDRESEEDCRNG`) that
+  2. Issue a `RNDRESEEDCRNG` system call.
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply docs/snapshotting/random-for-clones.md]: I have corrected the ioctl misnomer and restructured suggestions to still involve adding entropy before reseeding.

**Review by bchalios — APPROVED** [review]

**Review by dianpopa — APPROVED** [review]

**Pull request merged**